mm2src/coins/eth/web3_transport.rs

```diff
@@ -159,13 +174,39 @@ async fn send_request(
     const REQUEST_TIMEOUT_S: f64 = 60.;
 
     let mut errors = Vec::new();
-    for uri in uris.iter() {
-        let request = to_string(&request);
-        event_handlers.on_outgoing_request(request.as_bytes());
 
-        let mut req = http::Request::new(request.clone().into_bytes());
+    let serialized_request = to_string(&request);
+
+    for node in nodes.iter() {
+        let mut serialized_request = serialized_request.clone();
+        if node.gui_auth {
+            if let Some(generator) = gui_auth_validation_generator.clone() {
+                let mut json_payload: Json = serde_json::from_str(&serialized_request)?;
+                json_payload["signed_message"] = match EthCoin::generate_gui_auth_signed_validation(generator) {
```

Comment:
Adding a signature to each request is quite an overhead. I assumed it would be required only when a certain request count is reached, as follows:
1. If the client sends (relatively) low amount of requests, no balance validation is required.
2. If the client reaches a certain limit, it receives 429 code from auth proxy. It should submit a signed payload for "proof of funding".
3. If proof is verified, auth issues a JWT token to the client. The token has the number of requests (budget) available.
4. If budget is consumed too fast, the pubkey is banned.

I can see JWT usage in auth service, but the token seems to be included when request is sent to the origin (e.g. QuickNode) It's not issued for auth clients. Did I understand it correctly?

Reply:
We don't do anything with the pubkey yet. We recognize the clients by their IPs, not pubkey yet.

With the current way of proxy implementation, we hide the JWT from client-side. So there is no way anybody can try to crack our tokens because they never see them. The token is only exposed to the QuickNode from atomicdex-gui-auth proxy. I think signed messaage should be included to the each request, otherwise the proxy can  be manipulated(unless we start calculating requests based on pubkey). Also, people can use our proxy service for their own purposes by simply using curl.

I think caching the signed message for like 3-5 minutes would solve the overhead issue. So if you actively use the proxy service, you will create signed message 2-3 times at max in 10 minutes.

Reply:
> We don't do anything with the pubkey yet. We recognize the clients by their IPs, not pubkey yet.

So does it mean that someone can sign a message once and reuse it from multiple IPs, exceeding the total allowed RPS? The idea of "proof of funding" (and "trading history proof" in the future) allows distinguishing nodes that even use the same IP. So it's highly preferred that auth will track the number of requests by pubkeys too.

> I think caching the signed message for like 3-5 minutes would solve the overhead issue. So if you mm2 actively uses the proxy service, it will have to create signed message 2-3 times at max in 10 minutes.

I think it can be even significantly longer than 3-5 minutes. Basically, one signed message is enough to prove the ownership of a specific address.

There's also a problem with how the message itself. It's just the timestamp and auth simply checks `now > timestamp` https://github.com/KomodoPlatform/atomicdex-gui-auth/blob/main/main/src/security/sign.rs#L108. So I can sign e.g. `1974451550` and use it for 10 years from multiple IPs, not even needing to sign a new one. I mentioned in DM that message should be *randomly* generated by auth service. It can be pinned to a specific pubkey and also have a reasonable timeout, so it can't be reused forever.

Reply:
> So does it mean that someone can sign a message once and reuse it from multiple IPs, exceeding the total allowed RPS? The idea of "proof of funding" (and "trading history proof" in the future) allows distinguishing nodes that even use the same IP. So it's highly preferred that auth will track the number of requests by pubkeys too.

Rate limiting is generally known as limiting the requests by IPs, I implement the default idea. But in our case it's better to limit requests by client's pubkey. I will update that flow.

> So I can sign e.g. 1974451550 and use it for 10 years from multiple IPs, not even needing to sign a new one.

I can add another condition that checks if the message timestamp is longer than e.g 1 day or something. But this time they just need to generate signed message every 1 day. In this case, whatever we do people can debug and find a way to abuse our system. This would happen in every way of implementation. That's why we also have the **Status Checker** middleware. From audit logs, we can see if see someone is abusing our system and then update his/her status as `blocked` which will block off the incoming request from that address.

> that message should be randomly generated by auth service

I am not sure if I understand it correctly. If we do such thing then anyone can simply get their signed message and become valid.

> It can be pinned to a specific pubkey and also have a reasonable timeout, so it can't be reused forever.

Signed messagea are already pinned to specific pubkey. You can only generate signed message for your own address. Not for other ones. And the timeout defined from the client-side which I will give a limit on the auth-service side as well. e.g if you create signed message with timestamp of 10 years, it will not work.

Reply:
> I am not sure if I understand it correctly. If we do such thing then anyone can simply get their signed message and become valid.

I meant that auth will randomly generate the *payload* that client should then sign. And auth can also respond with the timeout when the signature of this payload will expire. But I need to think whether it's really required if additional timestamp check is added on auth side, will add one more comment tomorrow.

Reply:
rate limiting is updated to be `pubkey` based from `ip` based via [68cb411](https://github.com/KomodoPlatform/atomicdex-gui-auth/commit/68cb4118135fb1e4aa18124c6afb35457effe2ef)

Reply:
> Address based rate limiting should be fine right?

Yes, it's ok 🙂 

As per https://github.com/KomodoPlatform/atomicDEX-API/pull/1335#discussion_r929867837, it seems not required. Using this, we could implement the same functionality with signed message expiration fully controlled by auth server. However, signing the timestamp itself covers it in a simpler way. 

I just have a kind of "intuitive feeling" that if the client chooses expiration time, it might expose additional abuse vectors. But I don't have rational arguments and ideas on this, so we can leave the current implementation as is for now. Maybe I will figure it out later 🙂